token/transfer-hook-interface/src/instruction.rs

```diff
@@ -0,0 +1,206 @@
+//! Instruction types
+
+use {
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+        system_program,
+    },
+    spl_type_length_value::discriminator::{Discriminator, TlvDiscriminator},
+    std::convert::TryInto,
+};
+
+/// Instructions supported by the transfer hook interface.
+#[repr(C)]
+#[derive(Clone, Debug, PartialEq)]
+pub enum TransferHookInstruction {
+    /// Runs additional transfer logic.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[]` Source account
+    ///   1. `[]` Token mint
+    ///   2. `[]` Destination account
+    ///   3. `[]` Source account's owner/delegate
+    ///   4. `[]` Validation account
+    ///   5..5+M `[]` `M` additional accounts, written in validation account data
+    ///
+    Execute {
+        /// Amount of tokens to transfer
+        amount: u64,
+    },
+    /// Initializes the extra account metas on an account, writing into
+    /// the first open TLV space.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[w]` Account with extra account metas
+    ///   1. `[]` Mint
+    ///   2. `[s]` Mint authority
+    ///   3. `[]` System program
+    ///   4..4+M `[]` `M` additional accounts, to be written to validation data
+    ///
+    InitializeExtraAccountMetas,
+}
+/// TLV instruction type only used to define the discriminator. The actual data
+/// is entirely managed by `ExtraAccountMetas`, and it is the only data contained
+/// by this type.
+pub struct ExecuteInstruction;
+impl TlvDiscriminator for ExecuteInstruction {
+    const TLV_DISCRIMINATOR: Discriminator = Discriminator::new(EXECUTE_DISCRIMINATOR);
+}
+/// First 8 bytes of `hash::hashv(&["spl-transfer-hook-interface:execute"])`
+const EXECUTE_DISCRIMINATOR: [u8; Discriminator::LENGTH] = [105, 37, 101, 197, 75, 251, 102, 26];
+// annoying, but needed to perform a match on the value
+const EXECUTE_DISCRIMINATOR_SLICE: &[u8] = &EXECUTE_DISCRIMINATOR;
+/// First 8 bytes of `hash::hashv(&["spl-transfer-hook-interface:initialize-extra-account-metas"])`
+const INITIALIZE_EXTRA_ACCOUNT_METAS_DISCRIMINATOR: &[u8] = &[43, 34, 13, 49, 167, 88, 235, 235];
```

Comment:
I doubled checked both of these hashes here: https://emn178.github.io/online-tools/sha256.html